File: grind75/Jan14/230.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int {
    ret := -1 
    var dfs func(*TreeNode) bool
    dfs = func(root *TreeNode) bool {
        if root == nil {
            return false
        }
        if dfs(root.Left) {
            return true
        }
        k--
        if k == 0 {
            ret = root.Val
            return true
        }
        if dfs(root.Right) {
            return true
        }
        return false
    }
    dfs(root)
    return ret
}

File: grind75/Jan14/230.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> 
int:
        # inorder traversal iteratively
        stack = []
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            
            node = stack.pop()
            k -= 1
            if k == 0:
                return node.val
            if node.right:
                root = node.right
